Allow marking projects as "archived"

facutuesca opened the pull request:
## What is this?
Following the discussion [here](https://github.com/pypi/warehouse/issues/16844), this PR shows what the implementation could look like for adding "status markers" to projects.

**Specifically, this PR allows changing a project's status to "Archived"**.

## Details
The implementation is straightforward, since a `Project.lifecycle_status` field already exists (to mark a project as quarantined). This change adds a new "archived" value for that field, and adds the UI so that users can archive and unarchive their projects.

Even though the issue mentioned above proposes more statuses ("deprecated", "finished", etc), those can be ambiguous and overlapping, and would require further discussion to define clearly. "Archived", however, is already widely used on [GitHub](https://docs.github.com/en/repositories/archiving-a-github-repository/archiving-repositories) repos and its meaning should be easy for users to understand:
> You can archive a repository to make it read-only for all users and indicate that it's no longer actively maintained. You can also unarchive repositories that have been archived


## TODO
- [x] Actually restrict file uploads when project is Archived
- [x] Add admin UI to archive/unarchive projects

## Screenshots
### Project settings

<img width="700" alt="image" src="https://github.com/user-attachments/assets/b5065a9f-69df-4ce9-8c95-696c55aa2dae">

<img width="700" alt="image" src="https://github.com/user-attachments/assets/56f507b3-97a3-4e06-a449-c546fa7ed1ca">

<img width="700" alt="image" src="https://github.com/user-attachments/assets/25f5568f-4dbc-4ec6-9259-76b2eac518d7">



### Project landing page
<img width="700" alt="image" src="https://github.com/user-attachments/assets/ef764ddc-98a2-4c67-acd4-714fbef28db3">


cc @woodruffw @miketheman

miketheman reviewed (commented):
Generally, looking good, thanks for scoping it down to a single lifecycle status.

I think something good to add to the owner-facing management page would be a suggestion to publish a new release with a readme update or something to point folks where to go next.
Prior ideas have been to allow user-submitted text in the form, but I like that less, since then we have to figure out where to store that.

I think a warning modal popup confirmation might also be in order when archiving would be a good idea to prevent misclicks.

This also doesn't appear to block users from making further releases to an archived project - is that intentionally excluded? In my head, once Archived, it's in read-only mode - similar to GitHub's Archived.

miketheman commented:
I also just realized - this probably needs some added notation in the Admin UI/ability to change there as well. I didn't see anything in scope here yet.

facutuesca commented:
@miketheman 
> I think something good to add to the owner-facing management page would be a suggestion to publish a new release with a readme update or something to point folks where to go next.
Prior ideas have been to allow user-submitted text in the form, but I like that less, since then we have to figure out where to store that.
> 
> I think a warning modal popup confirmation might also be in order when archiving would be a good idea to prevent misclicks.> 

I changed the UI to add the suggestion and the modal popup. Since these are specific to archiving a project, I changed the UI so that instead of a dropdown with the backend statuses, the user sees two buttons to archive or unarchive their project (see the updated screenshots)

> This also doesn't appear to block users from making further releases to an archived project - is that intentionally excluded? In my head, once Archived, it's in read-only mode - similar to GitHub's Archived.

I'll add the restrictions once the rest is ready. I was thinking to only disallow uploads for now, WDYT?

> this probably needs some added notation in the Admin UI/ability to change there as well. I didn't see anything in scope here yet.

What should the UI for this look?

facutuesca commented:
Pushed a commit to disallow uploads on archived projects

facutuesca commented:
Added admin UI to archive/unarchive projects:


<img width="1637" alt="image" src="https://github.com/user-attachments/assets/22d02c90-42d5-40d5-848c-c09c4b364d8a">

miketheman commented on warehouse/utils/project.py:
```diff
@@ -189,3 +189,42 @@ def destroy_docs(project, request, flash=True):
+
+
+def archive_project(project: Project, request) -> None:
+    if (
+        project.lifecycle_status is None
+        or project.lifecycle_status == LifecycleStatus.QuarantineExit
+    ):
+        project.lifecycle_status = LifecycleStatus.Archived
+        project.record_event(
+            tag=EventTag.Project.ProjectArchiveEnter,
+            request=request,
+            additional={
+                "submitted_by": request.user.username,
+            },
+        )
+        request.session.flash("Project archived", queue="success")
+    else:
+        request.session.flash(
+            f"Cannot archive project with status {project.lifecycle_status}",
+            queue="error",
+        )
```
style: Consider early-exit inversion of the logic, like so (pseudocode):

```python
if not <conditions that must be met>:
    flash "cannot archive", error
    return

do the work
flash "done", success
```

thus communicate early on the exit clause, continue normally otherwise.
Removes one layer of indentation for "the work", and makes it simpler to create another if/else when necessary with fewer logical jumps to make.

Similar applies to `unarchive_project`

facutuesca replied:
fixed!

miketheman commented on warehouse/packaging/models.py:
```diff
@@ -330,7 +331,11 @@ def __acl__(self):
         # The project has zero or more OIDC publishers registered to it,
         # each of which serves as an identity with the ability to upload releases.
         for publisher in self.oidc_publishers:
-            acls.append((Allow, f"oidc:{publisher.id}", [Permissions.ProjectsUpload]))
+            if self.lifecycle_status != LifecycleStatus.Archived:
+                # Only allow uploads in non-archived projects
+                acls.append(
+                    (Allow, f"oidc:{publisher.id}", [Permissions.ProjectsUpload])
+                )
```
optimize: Adding the `if` will now happen on each iteration of the loop per `publisher`, but the project's lifecycle happens once, probably can shortcut the loop instead by placing the `if` outside.
Most projects only have a single publisher, but just in case, save the iteration if unnecessary.

facutuesca replied:
good catch, fixed!

miketheman commented on warehouse/packaging/models.py:
```diff
@@ -364,27 +369,27 @@ def __acl__(self):
+
+            if self.lifecycle_status == LifecycleStatus.Archived:
+                # Disallow upload permissions for archived projects
+                current_permissions = [
+                    p for p in current_permissions if p != Permissions.ProjectsUpload
+                ]
```
optimize: could this be instead?
```suggestion
                current_permissions.remove(Permissions.ProjectsUpload)
```
I _think_ each branch of `current_permissions` ends up with `ProjectsUpload`, and this might be faster than a list comprehension?

facutuesca replied:
makes sense!

miketheman reviewed (changes requested):
Coming along nicely, thanks for pushing forward on this. I had some notes inline, mostly structural in nature, but some questions as well - nothing too "heavy".

Let me know if anything isn't clear!

miketheman reviewed (changes requested):
Hey @facutuesca ! Nice progress here. I tested everything locally.

I left some more notes inline, let me know if they dpn't make sense!

facutuesca commented on warehouse/migrations/versions/12a43f12cc18_add_new_lifecycle_statuses.py:
```diff
@@ -0,0 +1,55 @@
+Add new lifecycle statuses
+
+Revision ID: 12a43f12cc18
+Revises: ed4cc2ef6b0f
```
Fixed

facutuesca commented on warehouse/templates/manage/project/settings.html:
```diff
@@ -40,6 +40,38 @@ <h2>{% trans %}Project settings{% endtrans %}</h2>
     <a href="{{ help_url }}"> (request an increase) </a>{% endtrans %}</li>
   </ul>
 
+  <h2>{% trans %}Archive project{% endtrans %}</h2>
```
good point, fixed:

<img width="400" alt="Screenshot 2025-01-16 at 16 35 58" src="https://github.com/user-attachments/assets/07f3abc0-f917-4d54-9bf9-3c1cb747d73f" />

facutuesca commented on warehouse/templates/manage/project/settings.html:
```diff
@@ -40,6 +40,38 @@ <h2>{% trans %}Project settings{% endtrans %}</h2>
+  <div class="callout-block callout-block--danger" data-controller="archive-confirm">
+  <p>
+    {% trans %}
+    Archiving a project will block any new uploads. Before doing so, we
```
fixed

facutuesca commented on warehouse/admin/templates/admin/projects/detail.html:
```diff
@@ -582,6 +582,33 @@ <h3 class="card-title">Prohibit Project Name</h3>
   </div>
 </form> <!-- .card #prohibitedProjectName -->
 
+<div class="card card-warning card-outline collapsed-card" id="archivedProjectName">
```
oh yeah that's a good idea, fixed:

<img width="600" alt="Screenshot 2025-01-16 at 16 53 00" src="https://github.com/user-attachments/assets/ff528d97-0f94-4012-862a-67893686cf45" />

miketheman commented on warehouse/migrations/versions/12a43f12cc18_add_new_lifecycle_statuses.py:
```diff
@@ -0,0 +1,55 @@
+Add new lifecycle statuses
+
+Revision ID: 12a43f12cc18
+Revises: e24aa37164e72
```
```suggestion
Revises: 24aa37164e72
```

miketheman commented on warehouse/templates/manage/project/settings.html:
```diff
@@ -351,6 +351,43 @@ <h3>
+  {% set can_be_unarchived = project.lifecycle_status == "archived" %}
+  <div class="callout-block callout-block--danger" data-controller="archive-confirm">
+  <p>
+    {% trans readme_description_href='https://packaging.python.org/en/latest/guides/making-a-pypi-friendly-readme/'%}
```
Drop locale and version to let the platform auto-route
```suggestion
    {% trans readme_description_href='https://packaging.python.org/guides/making-a-pypi-friendly-readme/'%}
```
